### PR TITLE
Remove the cast from the empty web project template

### PR DESCRIPTION
**Pull request by davidfowl**
- Remove the cast from the empty web project template

### CODE REVIEW

**Review by halter73 — APPROVED** [review]

File: src/ProjectTemplates/Web.ProjectTemplates/EmptyWeb-CSharp.csproj.in [hunk 1]
```diff
@@ -4,6 +4,7 @@
     <TargetFramework>${DefaultNetCoreTargetFramework}</TargetFramework>
     <NoDefaultLaunchSettingsFile Condition="'$(ExcludeLaunchSettings)' == 'True'">True</NoDefaultLaunchSettingsFile>
     <RootNamespace Condition="'$(name)' != '$(name{-VALUE-FORMS-}safe_namespace)'">Company.WebApplication1</RootNamespace>
+    <LangVersion>preview</LangVersion>
```
Comment (davidfowl) [inline src/ProjectTemplates/Web.ProjectTemplates/EmptyWeb-CSharp.csproj.in:7]: I *think* we'll be able to remove this when we get a new SDK.

cc @jaredpar

Comment (jaredpar) [reply src/ProjectTemplates/Web.ProjectTemplates/EmptyWeb-CSharp.csproj.in:7]: This will go away in .NET 6 P7.

**Comment by halter73** [comment]
It looks like `Templates.Test.EmptyWebTemplateTest.EmptyWebTemplateCSharp` is having issues again. `error CS1593: Delegate 'RequestDelegate' does not take 0 arguments` seems to imply the new MapGet overloads added in preview4 still not being tested again. Is this actually testing against the updated template and aspnet runtime @dotnet/aspnet-build?

```
Templates.Test.EmptyWebTemplateTest.EmptyWebTemplateCSharp
Project new web  failed to publish. Exit code 1.
/datadisks/disk1/work/A1E60924/p/dotnet-cli/dotnet publish --no-restore -c Release /bl
StdErr:
StdOut: Microsoft (R) Build Engine version 17.0.0-preview-21256-03+d07c47ade for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

/datadisks/disk1/work/A1E60924/p/dotnet-cli/sdk/6.0.100-preview.5.21264.3/MSBuild.dll -maxcpucount -property:Configuration=Release -target:Publish -verbosity:m /bl ./AspNet.bzpld23ioa3.csproj
You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
/datadisks/disk1/work/A1E60924/w/AC8C094C/e/Templates/BaseFolder/AspNet.bzpld23ioa3/Program.cs(13,17): error CS1593: Delegate 'RequestDelegate' does not take 0 arguments [/datadisks/disk1/work/A1E60924/w/AC8C094C/e/Templates/BaseFolder/AspNet.bzpld23ioa3/AspNet.bzpld23ioa3.csproj]

Expected: True
Actual:   False
```

**Comment by dougbu** [comment]
> more likely not testing with the preview compiler

@halter73 we should use the Microsoft.Net.Compilers.Toolset version from eng/Versions.props everywhere, even on Helix machines when testing templates (because they're special-cased). If that covers the complete "preview compiler", everything should be fine.

Unfortunately the Helix logs are far too detailed to be useful. The one interesting bit is
``` text
[xUnit.net 00:00:56.67]         | [47.544s] Templates.Test.EmptyWebTemplateTest Information: [ERROR] Reconciling library Microsoft.Net.Compilers.Toolset/4.0.0-2.21263.6
[xUnit.net 00:00:56.67]         | [47.544s] Templates.Test.EmptyWebTemplateTest Information: [ERROR] Library Microsoft.Net.Compilers.Toolset/4.0.0-2.21263.6 does not exist
```
This may indicate the SDK (or something) attempts to override https://github.com/dotnet/aspnetcore/blob/main/eng/Versions.props#L159 (which hasn't changed in a couple of months). I'm not sure what version it falls back to.

**Comment by davidfowl** [comment]
@dougbu we need to check in changes that will work with the latest VS and SDK before preview6 ships. There seems be an issue with ingestion of newer bits here and that's blocking this change. I don't really understand how any of this works can you help diagnose this please?

File: src/ProjectTemplates/Web.ProjectTemplates/EmptyWeb-CSharp.csproj.in (hunk 1 above)
Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/EmptyWeb-CSharp.csproj.in:7]: @jaredpar can you help me understand this better as the docs seem to say preview frameworks will always default to their matching preview language: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version

**Comment by davidfowl** [comment]
So I have a clue, when I build a sample in this repository with LangVersion set to 6.0 I see this in a binlog:

```
CommandLineArguments = C:\dev\git\AspNetCore\.dotnet\dotnet.exe exec "C:\Users\david\.nuget\packages\microsoft.net.compilers.toolset\3.10.0-1.final\tasks\netcoreapp3.1\bincore\csc.dll" /noconfig /unsafe- /checked- /nowarn:1701,1702,1705,NU5105,CS1591,RS0041,CA1416,1701,1702 /fullpaths /nostdlib+ /errorreport:prompt /warn:5 /define:TRACE;DEBUG;NET;NET6_0;NETCOREAPP;NET5_0_OR_GREATER;NET6_0_OR_GREATER;NETCOREAPP1_0_OR_GREATER;NETCOREAPP1_1_OR_GREATER;NETCOREAPP2_0_OR_GREATER;NETCOREAPP2_1_OR_GREATER;NETCOREAPP2_2_OR_GREATER;NETCOREAPP3_0_OR_GREATER;NETCOREAPP3_1_OR_GREATER /highentropyva+ ..
```

Any ideas @dougbu ? This might be an arcade question... The exact same code builds fine outside of this repo with the same outer command line.

**Comment by davidfowl** [comment]
I think it's coming from here https://github.com/dotnet/aspnetcore/blob/9fd548ff70ca8f64c5a41d3f1ddaddfb4ad7afd4/eng/Versions.props#L160

**Comment by davidfowl** [comment]
OK I figured it out. Lets see if it breaks the entire repo 😄

**Comment by davidfowl** [comment]
Feel free to let me know if this change is crazy because I removed some fundamental piece of infrastructure.

File: eng/tools/GenerateFiles/Directory.Build.targets.in [hunk 2]
```diff
@@ -54,10 +54,6 @@
           Condition=" '$(IsServicingBuild)' != 'true' ">${MicrosoftAspNetCoreAppRefVersion}</TargetingPackVersion>
     </KnownFrameworkReference>
 
-    <PackageReference Include="Microsoft.Net.Compilers.Toolset"
-      Version="${MicrosoftNetCompilersToolsetVersion}"
-      PrivateAssets="all"
-      IsImplicitlyDefined="true" />
```
Comment (halter73) [inline eng/tools/GenerateFiles/Directory.Build.targets.in:60]: Do we know why this was here before?

Comment (davidfowl) [reply eng/tools/GenerateFiles/Directory.Build.targets.in:60]: I don't know, we'll see if something breaks. I spoke to Doug and he's going to follow up and do some research.

Comment (dougbu) [reply eng/tools/GenerateFiles/Directory.Build.targets.in:60]: @wtgodbe introduced this long ago (for a 3.0.0 preview) in 23a6e3ead9a / #11818. We were reacting to a Roslyn breaking change and needed a newer toolset than Arcade (think we set `$(UsingToolMicrosoftNetCompilers)` at one point) or `msbuild` references by default. Maestro++ managed the version for a while but we stopped that at some point. @BrennanConroy and @pranavkm were among the few to change the version after we made the version manual.

I don't remember the details of the breaking change but it had something to do w/ nullability annotations and @roji did relevant work in the efcore repo. @jaredpar may know though he wasn't involved in the conversation back in the summer of 2019.

----

As a follow up, we should do one of

1. change the manually-controlled toolset version and re-introduce this package reference
2. clean up https://github.com/dotnet/aspnetcore/blob/8ee0b03d15fe5a3fbba86fbcf165bd8e5da53d6e/eng/Versions.props#L160
3. enable  `$(UsingToolMicrosoftNetCompilers)`, meaning Arcade updates will bump the toolset version on occasion

I don't know enough about the toolset to have an opinion on which choice is best. @jaredpar❔

Comment (jaredpar) [reply eng/tools/GenerateFiles/Directory.Build.targets.in:60]: In general repositories should avoid explicitly referencing this package and instead prefer the one that just comes with arcade. The arcade version is updated at a regular cadence (roughly every other week). The idea being that this update cadence means we're regularly pushing new versions of the compiler to our partners in .NET. 

This type of change where the version is pinned should be reserved for cases where a breaking change, bug, etc ... hits the repo and you need to pause the updates until we get a fix out. That appears to be what happened here except we forget to undo the pin once the bug was fixed.

Comment (dougbu) [reply eng/tools/GenerateFiles/Directory.Build.targets.in:60]: >  The arcade version is updated at a regular cadence (roughly every other week).

Right now, the Arcade version (see https://github.com/dotnet/arcade/blob/05634736a53ee65b0d19cf913d77d1f5d4293ebc/eng/Versions.props#L37) is far behind what the SDK uses by default (see https://github.com/dotnet/sdk/blob/ef6249b3b034b12f127b0d3a72ea428979f44413/eng/Versions.props#L120). Right now Arcade is only at `v3.10.0-4.21329.37` while the SDK is at `v4.0.0-2.21354.7`. I suggest we simply remove our confusing `$(MicrosoftNetCompilersToolsetVersion)` property and leave `$(UsingToolMicrosoftNetCompilers)` unset i.e. my option (2). Agreed @dotnet/aspnet-build @jaredpar❔

**Pull request merged**